omnibus/files/private-chef-cookbooks/private-chef/libraries/preflight_solr_validator.rb

```diff
@@ -16,48 +16,140 @@
+  attr_reader :cs_solr_attr, :cs_erchef_attr,
+              :cs_elasticsearch_attr, :cs_rabbitmq_attr,
+              :cs_opscode_expander_attr
+
+  # The node_*attr variables hold the default configuration
+  attr_reader :node_solr_attr, :node_elasticsearch_attr, :node_erchef_attr
 
   def initialize(node)
     super
+
     @cs_solr_attr = PrivateChef['opscode_solr4']
     @cs_erchef_attr = PrivateChef['opscode_erchef']
+    @cs_elasticsearch_attr = PrivateChef['elasticsearch']
+    @cs_rabbitmq_attr = PrivateChef['rabbitmq']
+    @cs_opscode_expander_attr = PrivateChef['opscode_expander']
+
     @node_erchef_attr = node['private_chef']['opscode-erchef']
     @node_solr_attr = node['private_chef']['opscode-solr4']
+    @node_elasticsearch_attr = node['private_chef']['elasticsearch']
   end
 
   def run!
-    verify_sane_reindex_sleep_times
+    verify_consistent_reindex_sleep_times
+
+    verify_one_search_provider
+    verify_es_disabled_if_user_set_external_solr
+    verify_unused_services_are_disabled_if_using_internal_es
+    verify_no_explicit_external_disable_when_es_enabled
+
     warn_unchanged_external_flag
     verify_external_url
     verify_erchef_config
   end
 
-  def verify_sane_reindex_sleep_times
+  def verify_consistent_reindex_sleep_times
     final_min = cs_erchef_attr['reindex_sleep_min_ms'] || node_erchef_attr['reindex_sleep_min_ms']
     final_max = cs_erchef_attr['reindex_sleep_max_ms'] || node_erchef_attr['reindex_sleep_max_ms']
     if final_min > final_max
-      fail_with <<~EOM
+      fail_with err_SOLR001_failed_validation(final_min, final_max)
+    end
+  end
 
-        opscode_erchef['reindex_sleep_min_ms'] (#{final_min}) is greater than
-        opscode_erchef['reindex_sleep_max_ms'] (#{final_max})
+  def verify_one_search_provider
+    if elasticsearch_enabled? && solr_enabled?
+      fail_with err_SOLR002_failed_validation
+    end
+  end
 
-        The maximum sleep time should be greater or equal to the minimum sleep
-        time.
-      EOM
+  # If an external search provider (ES or Solr) was explicitly enabled
+  # by the user, then we expect that both internal search services
+  # should be disabled.
+  def verify_es_disabled_if_user_set_external_solr
+    # NOTE(ssd) 2020-05-13: This might impact the people who we gave
+    # pre-release access to.
+    if cs_solr_attr['external'] && elasticsearch_enabled?
+      fail_with err_SOLR003_failed_validation
+    end
+
+    if solr_enabled? && external?
+      fail_with errl_SOLR004_failed_validation
+    end
+  end
+
+  # The user might have a combination of solr, rabbitmq or
+  # opscode-expander explicitly enabled in their config.
+  #
+  # If we are using the new internal elasticsearch, this is an
+  # unsupported configuration.
+  #
+  def verify_unused_services_are_disabled_if_using_internal_es
+    if elasticsearch_enabled?
+      msg = []
+      if cs_solr_attr['enable']
+        msg << err_SOLR005_failed_validation
```

Comment:
err_SOLR002_failed_validation will always catch this case; But would like to leave this here till the code stabilizes. (the migration PR is merged)